File: semester-2/programming/labs/zad8_battleship_game/Ships/Ship.cpp
//
// Created by marling on 6/6/23.
//

#include "Ship.hpp"



Ship::Ship(int length, const Coordinates& initialCoordinate)
        : length(length), coordinate(initialCoordinate) {}


int Ship::getLength() const {
    return length;
}


Coordinates Ship::getCoordinate() {
    return coordinate;
}

std::vector<Coordinates>& Ship::getCoordinates() {
    return coordinates;
}

void Ship::setCoordinate(Coordinates coord) {
    coordinate = coord;
}

void Ship::setCoordinates(const Coordinates& head, bool isHorizontal) {

    coordinates.emplace_back(head);

    int row = head.getRow();
    int column = head.getColumn();
    for (int i = 1; i < length; ++i) {
        if (isHorizontal) {
            coordinates.emplace_back(row, column + i);
        } else {
            coordinates.emplace_back(row - i, column);
        }
    }
}



Carrier::Carrier(int, const Coordinates& coord) : Ship(5, coord) {}

Battleship::Battleship(int, const Coordinates& coord) : Ship(4, coord) {}

Cruiser::Cruiser(int, const Coordinates& coord) : Ship(3, coord) {}

Destroyer::Destroyer(int, const Coordinates& coord) : Ship (2, coord) {}

Submarine::Submarine(int, const Coordinates& coord) : Ship(1, coord) {}

File: semester-3/Python/qt_battleship_game/game/ships/ship.py
class Ship:
    def __init__(self, length):
        self._length = length
        self._orientation = "vertical"
        self._is_set = False
        self._coordinates = []
        self._neighboring_coordinates = []

    @property
    def length(self):
        return self._length

    @property
    def orientation(self):
        return self._orientation

    @orientation.setter
    def orientation(self, value):
        self._orientation = value

    @property
    def is_set(self):
        return self._is_set

    @is_set.setter
    def is_set(self, value):
        self._is_set = value

    @property
    def coordinates(self):
        return self._coordinates

    @coordinates.setter
    def coordinates(self, value):
        self._coordinates = value

    @property
    def neighboring_coordinates(self):
        return self._neighboring_coordinates

    @neighboring_coordinates.setter
    def neighboring_coordinates(self, value):
        self._neighboring_coordinates = value

    def is_sunk(self, board):
        for coordinate in self._coordinates:
            if not board.get_cell(coordinate).has_been_hit:
                return False
        return True


class Submarine(Ship):
    def __init__(self):
        super().__init__(1)


class Destroyer(Ship):
    def __init__(self):
        super().__init__(2)


class Cruiser(Ship):
    def __init__(self):
        super().__init__(3)


class Battleship(Ship):
    def __init__(self):
        super().__init__(4)


class Carrier(Ship):
    def __init__(self):
        super().__init__(5)

